be/src/util/block_compression.cpp

```diff
@@ -442,6 +451,103 @@ class ZstdBlockCompression final : public BlockCompressionCodec {
     size_t max_compressed_len(size_t len) const override { return ZSTD_compressBound(len); }
 };
 
+class GzipBlockCompression final : public ZlibBlockCompression {
+public:
+    static const GzipBlockCompression* instance() {
+        static GzipBlockCompression s_instance;
+        return &s_instance;
+    }
+
+    ~GzipBlockCompression() override = default;
+
+    Status init_compress_stream(z_stream& zstrm) const override {
+        zstrm.zalloc = Z_NULL;
+        zstrm.zfree = Z_NULL;
+        zstrm.opaque = Z_NULL;
+        auto zres = deflateInit2(&zstrm, Z_DEFAULT_COMPRESSION, Z_DEFLATED, MAX_WBITS + 16, 8, Z_DEFAULT_STRATEGY);
+        if (zres != Z_OK) {
+            return Status::InvalidArgument(
+                    strings::Substitute("Fail to do ZLib stream compress, error=$0, res=$1", zError(zres), zres));
+        }
+        return Status::OK();
+    }
+
+    Status compress(const Slice& input, Slice* output) const override {
+        std::vector<Slice> orig_slices;
+        orig_slices.emplace_back(input);
+        auto zres = ZlibBlockCompression::compress(orig_slices, output);
+        if (!zres.ok()) {
+            return Status::InvalidArgument(
+                    strings::Substitute("Fail to do ZLib compress, error=$0", zres.get_error_msg()));
+        }
+        return Status::OK();
+    }
+
+    Status decompress(const Slice& input, Slice* output) const override {
+        z_stream z_strm = {nullptr};
+        z_strm.zalloc = Z_NULL;
+        z_strm.zfree = Z_NULL;
+        z_strm.opaque = Z_NULL;
+
+        int ret = inflateInit2(&z_strm, MAX_WBITS + 16);
+        if (ret < 0) {
+            std::stringstream ss;
+            ss << "Failed to init inflate. status code: " << ret;
+            return Status::InternalError(ss.str());
+        }
+
+        // 1. set input and output
+        z_strm.next_in = reinterpret_cast<Bytef*>(input.data);
+        z_strm.avail_in = input.size;
+        z_strm.next_out = reinterpret_cast<Bytef*>(output->data);
+        z_strm.avail_out = output->size;
+
+        while (z_strm.avail_out > 0 && z_strm.avail_in > 0) {
+            // inflate() performs one or both of the following actions:
+            //   Decompress more input starting at next_in and update next_in and avail_in
+            //       accordingly.
+            //   Provide more output starting at next_out and update next_out and avail_out
+            //       accordingly.
+            // inflate() returns Z_OK if some progress has been made (more input processed
+            // or more output produced)
+
+            int ret = inflate(&z_strm, Z_NO_FLUSH);
+
+            VLOG(10) << "gzip dec ret: " << ret;
+
+            if (ret == Z_BUF_ERROR) {
```

Comment:
According to the definition, this error looks like impossible to happen because we have already check both avail_out and avail_in must be greater than 0.

https://refspecs.linuxbase.org/LSB_3.0.0/LSB-Core-generic/LSB-Core-generic/zlib-inflate-1.html